pkg/flavor/kubernetes/multiwriter.go

```diff
@@ -0,0 +1,673 @@
+	}
+
+	// create pvc
+	newClaim, err := flavor.CreateNFSPVC(claimClone)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// create deployment with name hpe-multiwriter-<originalclaim-uid>
+	deploymentName := fmt.Sprintf("%s%s", deploymentNamePrefix, claim.ObjectMeta.UID)
+	err = flavor.CreateNFSDeployment(deploymentName, newClaim.ObjectMeta.Name)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// create service with name hpe-multiwriter-svc-<originalclaim-uid>
+	serviceName := fmt.Sprintf("%s%s", serviceNamePrefix, claim.ObjectMeta.UID)
+	err = flavor.CreateNFSService(serviceName)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// Get capacity requested
+	reqVolumeSize := int64(defaultVolumeSize) // Default value
+	if request.GetCapacityRange() != nil {
+		reqVolumeSize = request.GetCapacityRange().GetRequiredBytes()
+	}
+
+	// Return newly created volume
+	return &csi.Volume{
+		VolumeId:      fmt.Sprintf("%s", claim.ObjectMeta.UID),
+		CapacityBytes: reqVolumeSize,
+		VolumeContext: make(map[string]string),
+	}, false, nil
+}
+
+// DeleteMultiWriterVolume deletes multi-writer volume which represents nfs pvc, deployment and service
+func (flavor *Flavor) DeleteMultiWriterVolume(claimName string) error {
+	log.Tracef(">>>>> DeleteMultiWriterVolume with %s", claimName)
+	defer log.Tracef("<<<<< DeleteMultiWriterVolume")
+
+	// get pvc from request
+	claim, err := flavor.getClaimFromClaimName(claimName)
+	if err != nil {
+		return err
+	}
+	// delete service with name hpe-multiwriter-svc-<originalclaim-uid>
+	serviceName := fmt.Sprintf("%s%s", serviceNamePrefix, claim.ObjectMeta.UID)
+	err = flavor.DeleteNFSService(serviceName)
+	if err != nil {
+		log.Errorf("unable to delete nfs service %s as part of cleanup, err %s", serviceName, err.Error())
+	}
+
+	// delete deployment with name hpe-multiwriter-<originalclaim-uid>
+	deploymentName := fmt.Sprintf("%s%s", deploymentNamePrefix, claim.ObjectMeta.UID)
+	err = flavor.DeleteNFSDeployment(deploymentName)
+	if err != nil {
+		log.Errorf("unable to delete nfs deployment %s as part of cleanup, err %s", deploymentName, err.Error())
+	}
+	// delete pvc with name hpe-multiwriter-pvc-<originalclaim-uid>
+	// if deployment is still around, then pvc cannot be deleted due to protection, try to cleanup as best effort
+	pvcName := fmt.Sprintf("%s-%s", pvcNamePrefix, claim.ObjectMeta.UID)
+	err = flavor.DeleteNFSPVC(pvcName)
+	if err != nil {
+		log.Errorf("unable to delete nfs pvc %s as part of cleanup, err %s", pvcName, err.Error())
+	}
+	return err
+}
+
+func (flavor *Flavor) HandleMultiWriterNodePublish(req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	log.Tracef(">>>>> HandleMultiWriterNodePublish with volume %s target path %s", req.VolumeId, req.TargetPath)
+	defer log.Tracef("<<<<< HandleMultiWriterNodePublish")
+
+	// get service with matching volume-id(i.e original claim-id)
+	serviceName := fmt.Sprintf("%s%s", serviceNamePrefix, req.VolumeId)
```

Comment:
Note that we can also store whatever we want in the context of the request.  Those properties will propagate down here if we need to have a more elegant service name.

Reply:
got it..i just used volume-id to track it back to the original volume. so left publishContext empty, can you give examples of any properties we can use?